## Final Project Submission

Please fill out:
* **Student name:**                       Kinyanjui Chris Kamau
* **Student pace:**                       full time
* **Scheduled project review date/time:** 
* **Instructor name:**                    Antonny Muiko



# Data Understanding.

## Movie_budgets.csv.gz

In [1]:
#importing the relevant modules

#pandas
import pandas as pd

# matlab
import matplotlib.pyplot as plt

# to ensure all data visualizations appear on the same notebook and also stored here.
%matplotlib inline 

#seaborn
import seaborn as sns

#sqlite3
import sqlite3

#csv
import csv

#regex
import re

First file I shall load then explore is the ```'tn.movie_budgets.csv.gz'```.

In [2]:
# using pandas to load the file into movie_budget_db

movie_budget_df = pd.read_csv('zippedData/tn.movie_budgets.csv.gz')

movie_budget_df.head()

id  release_date                                        movie  \
0   1  Dec 18, 2009                                       Avatar   
1   2  May 20, 2011  Pirates of the Caribbean: On Stranger Tides   
2   3   Jun 7, 2019                                 Dark Phoenix   
3   4   May 1, 2015                      Avengers: Age of Ultron   
4   5  Dec 15, 2017            Star Wars Ep. VIII: The Last Jedi   

  production_budget domestic_gross worldwide_gross  
0      $425,000,000   $760,507,625  $2,776,345,279  
1      $410,600,000   $241,063,875  $1,045,663,875  
2      $350,000,000    $42,762,350    $149,762,350  
3      $330,600,000   $459,005,868  $1,403,013,963  
4      $317,000,000   $620,181,382  $1,316,721,747

In [3]:
#let us define some functions to get relevant information from the Data Frames.

#get information

def get_info(data):
    info_ = data.info()
    return info_
#get column names

def get_columns(data):
    columns=data.columns
    return columns


#type casting function for any column of numerical data stored as a string.
def int_conv(data,col_name):
    #first we remove the $ sign for the typecasting to work.
    data[col_name] = data[col_name].map(lambda col: re.sub('[^0-9]','', col))
    
    #then we convert/typecast.
    data[col_name] = data[col_name].astype(int)
    return data[col_name]

# defining a function to look for duplicates from the DataFrames.
def get_duplicate(data):
    duplicate = data.duplicated().value_counts()
    return duplicate


In [4]:
# getting info for the DataFrame "movie_budget_db"
get_info(movie_budget_df)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5782 entries, 0 to 5781
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   id                 5782 non-null   int64 
 1   release_date       5782 non-null   object
 2   movie              5782 non-null   object
 3   production_budget  5782 non-null   object
 4   domestic_gross     5782 non-null   object
 5   worldwide_gross    5782 non-null   object
dtypes: int64(1), object(5)
memory usage: 271.2+ KB


## Analysis.

The DataFrame has a total of 6 columns and 5782 rows.

From the information above it is clear it has no missing value.

On further investigation some data types are in string form which will make 
it hard to sort them and use them for statistical analysis.

The problem columns are ```'production_budget'```,```'domestic_gross'``` and ```'worldwide_gross'```.

In [5]:
#type casting the three columns
#production_budget
int_conv(movie_budget_df,'production_budget')

0       425000000
1       410600000
2       350000000
3       330600000
4       317000000
          ...    
5777         7000
5778         6000
5779         5000
5780         1400
5781         1100
Name: production_budget, Length: 5782, dtype: int64

In [6]:
#domestic_gross
int_conv(movie_budget_df,'domestic_gross')


0       760507625
1       241063875
2        42762350
3       459005868
4       620181382
          ...    
5777            0
5778        48482
5779         1338
5780            0
5781       181041
Name: domestic_gross, Length: 5782, dtype: int64

In [7]:
#worldwide_gross
int_conv(movie_budget_df,'worldwide_gross')

0       2776345279
1       1045663875
2        149762350
3       1403013963
4       1316721747
           ...    
5777             0
5778        240495
5779          1338
5780             0
5781        181041
Name: worldwide_gross, Length: 5782, dtype: int64

In [8]:
get_info(movie_budget_df)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5782 entries, 0 to 5781
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   id                 5782 non-null   int64 
 1   release_date       5782 non-null   object
 2   movie              5782 non-null   object
 3   production_budget  5782 non-null   int64 
 4   domestic_gross     5782 non-null   int64 
 5   worldwide_gross    5782 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 271.2+ KB


In [9]:
get_duplicate(movie_budget_df)

False    5782
dtype: int64

From the info above the Data Frame ```'movie_budget_df'``` doesnt have any duplicate rows.

## Reviews.tsv.gz

In [10]:
review_df = pd.read_csv('zippedData/rt.reviews.tsv.gz', delimiter = '\t', encoding = 'latin-1',compression='infer')
review_df.head()

id                                             review rating   fresh  \
0   3  A distinctly gallows take on contemporary fina...    3/5   fresh   
1   3  It's an allegory in search of a meaning that n...    NaN  rotten   
2   3  ... life lived in a bubble in financial dealin...    NaN   fresh   
3   3  Continuing along a line introduced in last yea...    NaN   fresh   
4   3             ... a perverse twist on neorealism...     NaN   fresh   

           critic  top_critic         publisher               date  
0      PJ Nabarro           0   Patrick Nabarro  November 10, 2018  
1  Annalee Newitz           0           io9.com       May 23, 2018  
2    Sean Axmaker           0  Stream on Demand    January 4, 2018  
3   Daniel Kasman           0              MUBI  November 16, 2017  
4             NaN           0      Cinema Scope   October 12, 2017

In [11]:
#getting info on review_df
get_info(review_df)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54432 entries, 0 to 54431
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   id          54432 non-null  int64 
 1   review      48869 non-null  object
 2   rating      40915 non-null  object
 3   fresh       54432 non-null  object
 4   critic      51710 non-null  object
 5   top_critic  54432 non-null  int64 
 6   publisher   54123 non-null  object
 7   date        54432 non-null  object
dtypes: int64(2), object(6)
memory usage: 3.3+ MB


In [12]:
#search for duplicate data
get_duplicate(review_df)

False    54423
True         9
dtype: int64

## movie_gross


In [13]:
gross_df = pd.read_csv('zippedData/bom.movie_gross.csv.gz', encoding = 'latin-1')
gross_df.head()

title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

  foreign_gross  year  
0     652000000  2010  
1     691300000  2010  
2     664300000  2010  
3     535700000  2010  
4     513900000  2010

In [14]:
get_info(gross_df)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3387 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3387 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3359 non-null   float64
 3   foreign_gross   2037 non-null   object 
 4   year            3387 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 132.4+ KB


In [15]:
get_duplicate(gross_df)

False    3387
dtype: int64

Missing values are obsereved. 

Also Some columns are string data types instead of integers.

No duplicates in the data.

## tmdb.movies.csv.gz

In [16]:
imdb_df = pd.read_csv('zippedData/tmdb.movies.csv.gz',delimiter= "," , encoding = 'latin-1')
imdb_df.head()

Unnamed: 0            genre_ids     id original_language  \
0           0      [12, 14, 10751]  12444                en   
1           1  [14, 12, 16, 10751]  10191                en   
2           2        [12, 28, 878]  10138                en   
3           3      [16, 35, 10751]    862                en   
4           4        [28, 878, 12]  27205                en   

                                 original_title  popularity release_date  \
0  Harry Potter and the Deathly Hallows: Part 1      33.533   2010-11-19   
1                      How to Train Your Dragon      28.734   2010-03-26   
2                                    Iron Man 2      28.515   2010-05-07   
3                                     Toy Story      28.005   1995-11-22   
4                                     Inception      27.920   2010-07-16   

                                          title  vote_average  vote_count  
0  Harry Potter and the Deathly Hallows: Part 1           7.7       10788  
1                      How to Train Your Dragon           7.7        7610  
2                                    Iron Man 2           6.8       12368  
3                                     Toy Story           7.9       10174  
4                                     Inception           8.3       22186

questions i need to answer if i want to understand this data set.

What are genre_ids?

popularity is determined by?

voting is out of?

In [17]:
get_info(imdb_df)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26517 entries, 0 to 26516
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         26517 non-null  int64  
 1   genre_ids          26517 non-null  object 
 2   id                 26517 non-null  int64  
 3   original_language  26517 non-null  object 
 4   original_title     26517 non-null  object 
 5   popularity         26517 non-null  float64
 6   release_date       26517 non-null  object 
 7   title              26517 non-null  object 
 8   vote_average       26517 non-null  float64
 9   vote_count         26517 non-null  int64  
dtypes: float64(2), int64(3), object(5)
memory usage: 2.0+ MB


No missing data.


In [18]:
get_duplicate(imdb_df)

False    26517
dtype: int64

In [19]:
#no duplicates observed
!ls

CONTRIBUTING.md  im.db	     movie_data_erd.jpeg  README.md
data.sqlite	 LICENSE.md  Movie_Project.ipynb  zippedData


## im.db.... SQL

In [22]:
conn = sqlite3.connect('zippedData/im.db')

In [23]:
cursor_ = conn.cursor()

cursor_.execute("""
SELECT name
FROM sqlite_master
WHERE type ='table';
""")

table_names = cursor_.fetchall()
table_names

[('movie_basics',),
 ('directors',),
 ('known_for',),
 ('movie_akas',),
 ('movie_ratings',),
 ('persons',),
 ('principals',),
 ('writers',)]